CDAP-7271 Fallback strategy for User Service Routing

gokulavasan opened the pull request:
JIRA : https://issues.cask.co/browse/CDAP-7271

Build : http://builds.cask.co/browse/CDAP-RUT198

gokulavasan committed:
CDAP-7271 Default Routing Strategy for User Services

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -1549,6 +1549,16 @@
+    <value>random</value>
+    <description>
+      If RouteConfig is not found for a particular user service, this property can be used to define the fallback
+      routing strategy. Allowed options are - random, smallest, largest, drop (smallest/largest is based on the string
```
What's the behavior of "drop"?

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/RouteFallbackStrategy.java:
```diff
@@ -0,0 +1,39 @@
+  LARGEST;
+
+  public static RouteFallbackStrategy valueOfRouteFallbackStrategy(String strategy) {
+    strategy = Strings.isNullOrEmpty(strategy) ? "" : strategy;
```
You can simply do:

``` java
try {
  return RouteFallbackStrategy.valueOf(strategy.toUpperCase());
} catch (Exception e) {
  return RouteFallbackStrategy.RANDOM;
}
```

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java:
```diff
@@ -39,48 +39,71 @@
 
   public UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId,
-                                     @Nullable String version) {
+                                     String fallbackStrategy, @Nullable String version) {
```
Better take a `RouteFallbackStrategy`. Turning a string into that enum is not the responsibility of this class.

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java:
```diff
@@ -39,48 +39,71 @@
-    this(serviceDiscovered, routeStore, serviceId, null);
+  @VisibleForTesting
+  UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId) {
+    this(serviceDiscovered, routeStore, serviceId, null, null);
```
Instead of passing in `null` for fallback strategy, better pass in the default (which in RANDOM).

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java:
```diff
@@ -39,48 +39,71 @@
+    String minMaxVersion = null;
     // If version is not null, then don't fetch any routeConfig.
     RouteConfig routeConfig = version == null ? routeStore.fetch(serviceId) : null;
+    boolean routeConfigUnAvailable = (routeConfig == null) || routeConfig.getRoutes().isEmpty();
```
Usually a positive boolean variable is more preferred. E.g. in this case, better have `hasRouteConfig` instead. This helps the logic easier to read. Consider `if (hasRouteConfig)` vs `if (!routeConfigUnAvailable)`. Double negative is harder to understand.

gokulavasan commented:
@chtyim Addressed comments and also simplified the logic in UserServiceEndpointStrategy. Will add more info about maintaining the IdentityHashMap in the JIRA to improve performance of the UserServiceEndpointStrategy.

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/VersionFilteredServiceDiscovered.java:
```diff
@@ -0,0 +1,71 @@
+import javax.annotation.Nullable;
+
+/**
+ * Return {@link Discoverable}s that match the version provided.
```
Actually `version` is optional and the matching is only done when version is provided.

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java:
```diff
@@ -39,48 +39,77 @@
+    }
+
+    RouteConfig routeConfig = routeStore.fetch(serviceId);
+    boolean routeConfigUnAvailable = (routeConfig == null) || routeConfig.getRoutes().isEmpty();
```
I commented before. Better flip the boolean and call it `hasRouteConfig`

gokulavasan replied:
I thought about it but all the if conditions need routeConfigUnAvailable. Otherwise it would be if(!hasRouteConfig). I can flip it if you think it will make it more readable.

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java:
```diff
@@ -39,48 +39,77 @@
+      }
+
+      if (fallbackStrategy.equals(RouteFallbackStrategy.RANDOM)) {
+        // Since version is null, we will do pull random pick across all discoverables
```
This comment is not accurate. Whether version is null or not, we are picking randomly. Just that if version is provided, we only pick among those that matches with the version.

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java:
```diff
@@ -89,4 +118,18 @@ public Discoverable pick() {
     return result;
   }
+
+  private double fetchWeight(boolean routeConfigUnavailable, String version, RouteConfig routeConfig) {
```
version should be `@Nullable`

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java:
```diff
@@ -89,4 +118,18 @@ public Discoverable pick() {
   }
+
+  private double fetchWeight(boolean routeConfigUnavailable, String version, RouteConfig routeConfig) {
+    if (routeConfigUnavailable || Strings.isNullOrEmpty(version)) {
```
To be consistent with the rest of the logic, should just check if `version == null`.

chtyim commented:
Only couple minor comments left. Rest LGTM.

gokulavasan commented:
@chtyim Thanks for the review Terence. Addressed the recent comments. Please take a look again.

gokulavasan commented:
@chtyim Added isValid method to RouteConfig and removed the boolean variable. Please take a look again. Thanks!

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/RouteConfigHttpHandler.java:
```diff
@@ -32,6 +30,7 @@
 import com.google.gson.reflect.TypeToken;
 import com.google.inject.Inject;
 import com.google.inject.Singleton;
+import org.apache.commons.math3.analysis.function.Add;
```
What is this?

gokulavasan commented:
@chtyim Removed the unused import. Please take a look again. Thanks!

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/route/store/RouteConfig.java:
```diff
@@ -35,4 +35,14 @@ public RouteConfig(Map<String, Integer> routes) {
+
+  public boolean isValid() {
+    int percentageSum = 0;
+    if (routes != null) {
```
Actually should just check if the routes is empty.

chtyim replied:
Always going through all routes can be time consuming see this is called within a loop.

gokulavasan commented:
@chtyim Made the change to compute isValid only once in the constructor of RouteConfig.

gokulavasan commented:
Got an LGTM. Merging since the build passed.
